extensions/amp-ad-network-doubleclick-impl/0.1/test/test-doubleclick-sra.js

```diff
@@ -0,0 +1,471 @@
+          cookieOptOut: 1,
+          categoryExclusions: 'food',
+          targeting: {hello: 'world'},
+        };
+        targeting2[TFCD] = 'some_other_tfcd';
+        const config2 = {
+          type: 'doubleclick',
+          height: 300,
+          width: 250,
+          'data-slot': '/1234/def/xyz',
+          'json': JSON.stringify(targeting2),
+        };
+        const element2 =
+        createElementWithAttributes(doc, 'amp-ad', config2);
+        const impl2 = new AmpAdNetworkDoubleclickImpl(element2);
+        sandbox.stub(impl2, 'generateAdKey_').withArgs(
+            '250x300').returns('2468');
+        element2.setAttribute(EXPERIMENT_ATTRIBUTE, MANUAL_EXPERIMENT_ID);
+        impl2.populateAdUrlState();
+        expect(constructSRABlockParameters([impl1, impl2])).to.jsonEqual({
+          'iu_parts': '1234,abc,def,xyz',
+          'enc_prev_ius': '0/1/2,0/2/3',
+          adks: '13579,2468',
+          'prev_iu_szs': '50x320,250x300',
+          'prev_scp':
+          'foo=bar&names=x,y,z&excl_cat=sports|hello=world&excl_cat=food',
+          co: '1',
+          adtest: 'on',
+          tfcd: 'some_tfcd',
+          eid: MANUAL_EXPERIMENT_ID,
+        });
+      });
+    });
+
+    describe('#initiateSraRequests', () => {
+      let xhrMock;
+
+      function createA4aSraInstance(networkId) {
+        const element =
+        createElementWithAttributes(doc, 'amp-ad', {
+          type: 'doubleclick',
+          height: 320,
+          width: 50,
+          'data-slot': `/${networkId}/abc/def`,
+        });
+        element.getAmpDoc = () => {
+          const ampdocService = Services.ampdocServiceFor(doc.defaultView);
+          return ampdocService.getAmpDoc(element);
+        };
+        element.isBuilt = () => {return true;};
+        element.getLayoutBox = () => {
+          return layoutRectLtwh(0, 0, 200, 50);
+        };
+        element.getPageLayoutBox = () => {
+          return element.getLayoutBox.apply(element, arguments);
+        };
+        element.getIntersectionChangeEntry = () => {return null;};
+        doc.body.appendChild(element);
+        const impl = new AmpAdNetworkDoubleclickImpl(element);
+        impl.useSra = true;
+        return impl;
+      }
+
+      function generateSraXhrMockCall(
+        validInstances, networkId, responses, opt_xhrFail, opt_allInvalid) {
+        dev().assert(validInstances.length > 1);
+        dev().assert(!(opt_xhrFail && opt_allInvalid));
+      // Start with nameframe method, SRA will override to use safeframe.
+        const headers = {};
+        headers[RENDERING_TYPE_HEADER] = XORIGIN_MODE.NAMEFRAME;
+      // Assume all implementations have same data slot.
+        const iuParts = encodeURIComponent(
+            validInstances[0].element.getAttribute('data-slot').split(/\//)
+        .splice(1).join());
+        const xhrWithArgs = xhrMock.withArgs(
+            sinon.match(
+                new RegExp('^https:\/\/securepubads\\.g\\.doubleclick\\.net' +
```

Comment:
Consider creating a helper function somewhere for quoting strings to be used in a regexp:

function escapeRegExp(string){
  return string.replace(/([.*+?^=!:${}()|\[\]\/\\])/g, "\\$1");
}